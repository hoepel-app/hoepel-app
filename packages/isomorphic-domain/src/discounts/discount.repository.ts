import { Observable } from 'rxjs'
import { Discounts } from './discounts'

export type DiscountsRepository = {
  getForTenant(tenantId: string): Observable<Discounts>
  put(entity: Discounts): Promise<void>
}
