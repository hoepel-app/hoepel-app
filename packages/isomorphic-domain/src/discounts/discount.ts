import { Price } from '@hoepel.app/types'

type AbsoluteDiscountValue = {
  kind: 'relative-discount'
  discountPercent: number
}
type RelativeDiscountValue = {
  kind: 'absolute-discount'
  discountCents: number
}
type DiscountValue = AbsoluteDiscountValue | RelativeDiscountValue

export type DiscountProps = {
  name: string
  discount: DiscountValue
}

export class Discount {
  private constructor(private readonly props: DiscountProps) {}

  toProps(): DiscountProps {
    return this.props
  }

  static fromProps(props: DiscountProps): Discount {
    return new Discount(props)
  }

  static createRelativeDiscount(name: string, percent: number): Discount {
    const discountPercent = Math.round(Math.min(100, Math.max(percent, 0)))

    return new Discount({
      name,
      discount: {
        kind: 'relative-discount',
        discountPercent: isNaN(discountPercent) ? 0 : discountPercent,
      },
    })
  }

  static createAbsoluteDiscount(name: string, price: Price): Discount {
    return new Discount({
      name,
      discount: {
        kind: 'absolute-discount',
        discountCents: price.totalCents,
      },
    })
  }

  /**
   * Apply a list of discounts to a price. The discounts will be applied in the array order.
   *
   * The result will never be negative (always at least 0.00EUR)
   *
   * @param to The price to apply these discounts to
   * @param discounts The list of discounts to be applied
   */
  static applyAllDiscounts(to: Price, discounts: readonly Discount[]): Price {
    return discounts.reduce((acc, current) => {
      return current.applyTo(acc)
    }, to)
  }

  /**
   * Applies this discount to a price.
   * The result will never be negative (always at least 0.00EUR)
   */
  applyTo(price: Price): Price {
    switch (this.props.discount.kind) {
      case 'absolute-discount': {
        return price.subtract(this.absoluteDiscount)
      }
      case 'relative-discount': {
        const pct = this.discountPercentage * 0.01
        return price.multiply(1 - pct)
      }
    }
  }

  get isRelativeDiscount(): boolean {
    return this.props.discount.kind === 'relative-discount'
  }

  get isAbsoluteDiscount(): boolean {
    return this.props.discount.kind === 'absolute-discount'
  }

  /**
   * Return the discount as a percentage ([0,100]).
   * If this discount is an absolute discount, returns 0%
   */
  get discountPercentage(): number {
    switch (this.props.discount.kind) {
      case 'absolute-discount':
        return 0
      case 'relative-discount':
        return this.props.discount.discountPercent
    }
  }

  /**
   * Return the absolute discount as a price to be substracted
   * If this discount is a relative discount, returns 0.00EUR
   */
  get absoluteDiscount(): Price {
    switch (this.props.discount.kind) {
      case 'absolute-discount':
        return Price.fromCents(this.props.discount.discountCents)
      case 'relative-discount':
        return Price.zero
    }
  }

  get name(): string {
    return this.props.name
  }

  withName(name: string): Discount {
    return Discount.fromProps({
      ...this.toProps(),
      name,
    })
  }
}
