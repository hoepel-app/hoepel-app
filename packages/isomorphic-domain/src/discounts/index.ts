export { AddDiscountCommand } from './commands/add-discount.command'
export { RemoveDiscountCommand } from './commands/remove-discount.command'
export { RenameDiscountCommand } from './commands/rename-discount.command'
export { DiscountsRepository } from './discount.repository'
export { Discounts, DiscountsProps } from './discounts'
export { Discount, DiscountProps } from './discount'
export { DiscountsApplicationService } from './discounts.application-service'
