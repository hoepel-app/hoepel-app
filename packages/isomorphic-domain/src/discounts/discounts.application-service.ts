import { DiscountsRepository } from './discount.repository'
import { Discounts } from './discounts'
import { Observable } from 'rxjs'
import { AddDiscountCommand } from './commands/add-discount.command'
import { CommandResult } from '@hoepel.app/ddd-library'
import { first } from 'rxjs/operators'
import { RenameDiscountCommand } from './commands/rename-discount.command'
import { RemoveDiscountCommand } from './commands/remove-discount.command'

export class DiscountsApplicationService {
  constructor(private readonly repo: DiscountsRepository) {}

  findDiscounts(tenantId: string): Observable<Discounts> {
    return this.repo.getForTenant(tenantId)
  }

  async addDiscount(command: AddDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (command.discount.name.trim() === '') {
      return {
        status: 'rejected',
        reason: 'Discount name is empty',
      }
    }

    if (!discounts.mayAddDiscount(command.discount)) {
      return {
        status: 'rejected',
        reason: `May not add the discount with name ${command.discount.name}`,
      }
    }

    await this.repo.put(discounts.withDiscountAdded(command.discount))

    return { status: 'accepted' }
  }

  async renameDiscount(command: RenameDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (command.newName.trim() === '') {
      return {
        status: 'rejected',
        reason: `New name is empty`,
      }
    }

    if (!discounts.discountWithNameExists(command.oldName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.oldName} does not exist`,
      }
    }

    if (discounts.discountWithNameExists(command.newName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.newName} already exists`,
      }
    }

    await this.repo.put(
      discounts.withDiscountRenamed(command.oldName, command.newName)
    )

    return { status: 'accepted' }
  }

  async removeDiscount(command: RemoveDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!discounts.discountWithNameExists(command.discountName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.discountName} does not exist`,
      }
    }

    await this.repo.put(discounts.withDiscountRemoved(command.discountName))

    return { status: 'accepted' }
  }
}
