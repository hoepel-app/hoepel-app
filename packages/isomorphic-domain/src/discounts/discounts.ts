import { Aggregate } from '@hoepel.app/ddd-library'
import { DiscountProps, Discount } from './discount'

export type DiscountsProps = {
  readonly tenantId: string
  readonly discounts: DiscountProps[]
}

export class Discounts implements Aggregate {
  private constructor(private readonly props: DiscountsProps) {}

  static createEmpty(tenantId: string): Discounts {
    return new Discounts({ tenantId, discounts: [] })
  }

  get id(): string {
    return `${this.tenantId}-discounts`
  }

  get tenantId(): string {
    return this.props.tenantId
  }

  get names(): ReadonlySet<string> {
    return new Set(this.discounts.map((discount) => discount.name))
  }

  get namesSorted(): readonly string[] {
    return [...this.names].sort()
  }

  get discounts(): readonly Discount[] {
    return [
      ...this.props.discounts.map((props) => Discount.fromProps(props)),
    ].sort((a, b) => a.name.localeCompare(b.name))
  }

  toProps(): DiscountsProps {
    return this.props
  }

  static fromProps(props: DiscountsProps): Discounts {
    return new Discounts(props)
  }

  get isEmpty(): boolean {
    return this.props.discounts.length === 0
  }

  mayAddDiscount(discount: Discount): boolean {
    return !this.discountWithNameExists(discount.name)
  }

  discountWithNameExists(name: string): boolean {
    return this.names.has(name)
  }

  findDiscountWithName(name: string): Discount | null {
    return this.discounts.find((discount) => discount.name === name) ?? null
  }

  withDiscountAdded(discount: Discount): Discounts {
    if (!this.mayAddDiscount(discount)) {
      return this
    }

    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts.map((d) => d.toProps()),
        discount.toProps(),
      ],
    })
  }

  withDiscountRenamed(oldName: string, newName: string): Discounts {
    const oldDiscount = this.findDiscountWithName(oldName)

    if (oldDiscount == null || this.discountWithNameExists(newName)) {
      return this
    }

    const renamedDiscount = oldDiscount.withName(newName)

    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts
          .filter((d) => d.name !== oldName)
          .map((d) => d.toProps()),
        renamedDiscount.toProps(),
      ],
    })
  }

  withDiscountRemoved(name: string): Discounts {
    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts
          .filter((d) => d.name !== name)
          .map((d) => d.toProps()),
      ],
    })
  }
}
