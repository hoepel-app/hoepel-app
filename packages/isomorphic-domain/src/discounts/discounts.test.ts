import { Discounts } from './discounts'
import { Discount } from './discount'
import { Price } from '@hoepel.app/types'

describe('Discounts', () => {
  const exampleDiscounts = Discounts.createEmpty('my-tenant-id')
    .withDiscountAdded(
      Discount.createAbsoluteDiscount('1 euro korting', Price.fromCents(100))
    )
    .withDiscountAdded(Discount.createRelativeDiscount('Kansentarief', 60))
    .withDiscountAdded(Discount.createRelativeDiscount('Tweede kind', 50))

  const emptyDiscounts = Discounts.createEmpty('my-tenant-id-here')

  test('serializing then deserializing yields same object', () => {
    const deserialized = Discounts.fromProps(
      JSON.parse(JSON.stringify(exampleDiscounts.toProps()))
    )

    expect(deserialized).toEqual(exampleDiscounts)
  })

  test('id', () => {
    expect(exampleDiscounts.id).toEqual('my-tenant-id-discounts')
  })

  describe('names', () => {
    it('returns names for discounts', () => {
      expect(exampleDiscounts.names).toEqual(
        new Set(['1 euro korting', 'Tweede kind', 'Kansentarief'])
      )
    })

    it('returns empty set when no discounts added', () => {
      expect(emptyDiscounts.names).toEqual(new Set())
    })
  })

  describe('namesSorted', () => {
    it('returns the name sorted alpabetically', () => {
      expect(exampleDiscounts.namesSorted).toEqual([
        '1 euro korting',
        'Kansentarief',
        'Tweede kind',
      ])
    })
  })

  describe('mayAddDiscount', () => {
    it('returns false if a discount with the same name already exists', () => {
      expect(
        exampleDiscounts.mayAddDiscount(
          Discount.createAbsoluteDiscount('Tweede kind', Price.fromCents(200))
        )
      ).toEqual(false)
    })
  })

  describe('discountWithNameExists', () => {
    it('true when discount with name exists', () => {
      expect(exampleDiscounts.discountWithNameExists('Kansentarief')).toEqual(
        true
      )
    })

    it('false when discount with name does not exists', () => {
      expect(exampleDiscounts.discountWithNameExists('Kansentarfie')).toEqual(
        false
      )
    })
  })

  describe('isEmpty', () => {
    it('true when no discounts', () => {
      expect(emptyDiscounts.isEmpty).toEqual(true)
    })

    it('false when discounts added', () => {
      expect(exampleDiscounts.isEmpty).toEqual(false)
    })
  })

  describe('withDiscountAdded', () => {
    it('does nothing when discount with same name is added', () => {
      expect(
        exampleDiscounts.withDiscountAdded(
          Discount.createAbsoluteDiscount('Kansentarief', Price.zero)
        )
      ).toEqual(exampleDiscounts)
    })
  })

  describe('withDiscountRenamed', () => {
    it('renames a discount', () => {
      expect(
        exampleDiscounts.withDiscountRenamed('Tweede kind', 'Derde kind')
      ).toMatchSnapshot()
    })

    it('does nothing when old name does not exist', () => {
      expect(exampleDiscounts.withDiscountRenamed('aoeu', 'Test')).toEqual(
        exampleDiscounts
      )
    })

    it('does nothing when new name already taken', () => {
      expect(
        exampleDiscounts.withDiscountRenamed('Tweede kind', '1 euro korting')
      ).toEqual(exampleDiscounts)
    })
  })

  describe('withDiscountRemoved', () => {
    it('does nothing with name is not added', () => {
      expect(exampleDiscounts.withDiscountRemoved('nope')).toEqual(
        exampleDiscounts
      )
    })

    it('removes discount', () => {
      expect(
        exampleDiscounts.withDiscountRemoved('Tweede kind')
      ).toMatchSnapshot()
    })
  })
})
