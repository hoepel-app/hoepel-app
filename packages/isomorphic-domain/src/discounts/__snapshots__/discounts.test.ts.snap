// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discounts withDiscountRemoved removes discount 1`] = `
Discounts {
  "props": Object {
    "discounts": Array [
      Object {
        "discount": Object {
          "discountCents": 100,
          "kind": "absolute-discount",
        },
        "name": "1 euro korting",
      },
      Object {
        "discount": Object {
          "discountPercent": 60,
          "kind": "relative-discount",
        },
        "name": "Kansentarief",
      },
    ],
    "tenantId": "my-tenant-id",
  },
}
`;

exports[`Discounts withDiscountRenamed renames a discount 1`] = `
Discounts {
  "props": Object {
    "discounts": Array [
      Object {
        "discount": Object {
          "discountCents": 100,
          "kind": "absolute-discount",
        },
        "name": "1 euro korting",
      },
      Object {
        "discount": Object {
          "discountPercent": 60,
          "kind": "relative-discount",
        },
        "name": "Kansentarief",
      },
      Object {
        "discount": Object {
          "discountPercent": 50,
          "kind": "relative-discount",
        },
        "name": "Derde kind",
      },
    ],
    "tenantId": "my-tenant-id",
  },
}
`;
